python/oneflow/multiprocessing/reductions.py

```diff
@@ -31,32 +34,54 @@
     pass
 
 
-def rebuild_tensor(cls, tensor_data, requires_grad):
-    t = flow.tensor(tensor_data)
-    if cls == Parameter:
-        # we have to pass requires_grad into constructor, rather than set it as an
-        # attribute later, because it's an important check for Integer Tensors to
-        # have requires_grad=False (or else they raise an error)
-        t = Parameter(t, requires_grad=requires_grad)
-    else:
-        t.requires_grad = requires_grad
+def rebuild_shm_tensor(shm, shape, dtype, requires_grad):
+    def delete_shm():
+        shm.close()
+        shm.unlink()
+
+    arr = np.ndarray(shape, dtype=dtype, buffer=shm.buf)
+    t = flow.from_numpy(arr)
+    t._register_storage_delete_hook(delete_shm)
```

Comment:
```suggestion
    t._register_storage_delete_hook(delete_shm, shm)
```

这里应该把 shm 传进来让 Tensor 维持它的生命周期吧？ @lixinqi

Reply:
如果直接传 std::function 的话，这个 function 中用到的对象可以在 pybind 做 py::function -> std::function 时自动捕获吗？

Reply:
我猜测pybind11肯定做了，否则就是一个非常明显的BUG。我很快做个实验。

Reply:
```bash
[lixinqi@oneflow-16 pybind11]$ cat example.cpp
#include <pybind11/pybind11.h>
#include <pybind11/functional.h>
#include <stdlib.h>

void Call(const std::function<void()>& Func) {
  Func();
}

PYBIND11_MODULE(example, m) {
    m.doc() = "pybind11 example plugin"; // optional module docstring

    m.def("Call", &Call);
}
[lixinqi@oneflow-16 pybind11]$ g++ -O3 -Wall -shared -std=c++11 -fPIC $(python3 -m pybind11 --includes) example.cpp -o example$(python3-config --extension-suffix)
[lixinqi@oneflow-16 pybind11]$ python3
Python 3.6.8 (default, Apr  2 2020, 13:34:55)
[GCC 4.8.5 20150623 (Red Hat 4.8.5-39)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import example
>>> x = [1, 2, 3, 4]
>>> def Print():
...     print(x)
...
>>> example.Call(Print)
[1, 2, 3, 4]
>>>
```
没问题，放心用。

Reply:
嗯嗯，我这里也实验了一下，看了下引用计数是正确的，我改一下

Reply:
下边的例子说服力更强。
```bash
[lixinqi@oneflow-16 pybind11]$ python3
Python 3.6.8 (default, Apr  2 2020, 13:34:55)
[GCC 4.8.5 20150623 (Red Hat 4.8.5-39)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> def Print():
...     x = [1, 2, 3, 4]
...     return lambda: print(x)
...
>>> import example
>>> example.Call(Print())
[1, 2, 3, 4]
>>>
```